vra/resource_machine.go

```diff
@@ -294,49 +295,203 @@ func resourceMachineUpdate(d *schema.ResourceData, m interface{}) error {
 		}
 
+		// ignore if the disk is already in detached state
+		if deleteMachineDiskNoContent != nil {
+			continue
+		}
+
 		stateChangeFunc := resource.StateChangeConf{
 			Delay:      5 * time.Second,
 			Pending:    []string{models.RequestTrackerStatusINPROGRESS},
-			Refresh:    machineStateRefreshFunc(*apiClient, *resizeMachine.Payload.ID),
+			Refresh:    machineStateRefreshFunc(*apiClient, *deleteMachineDiskAccepted.Payload.ID),
 			Target:     []string{models.RequestTrackerStatusFINISHED},
-			Timeout:    d.Timeout(schema.TimeoutUpdate),
+			Timeout:    d.Timeout(schema.TimeoutCreate),
 			MinTimeout: 5 * time.Second,
 		}
 
-		resourceIds, err := stateChangeFunc.WaitForState()
-		if err != nil {
-			return err
+		_, e := stateChangeFunc.WaitForState()
+		if e != nil {
+			return e
 		}
+	}
+
+	// get all the disks currently attached to the machine
+	getMachineDisksOk, err := apiClient.Disk.GetMachineDisks(disk.NewGetMachineDisksParams().WithID(id))
+	if err != nil {
+		return err
+	}
 
-		machineIds := resourceIds.([]string)
-		d.SetId(machineIds[0])
-		log.Printf("Finished to resize vra_machine resource with name %s", d.Get("name"))
+	diskIds := make([]string, len(getMachineDisksOk.GetPayload().Content))
+
+	for i, blockDevice := range getMachineDisksOk.GetPayload().Content {
+		diskIds[i] = *blockDevice.ID
 	}
-	log.Printf("Finished updating the vra_machine resource with name %s", d.Get("name"))
 
-	return resourceMachineRead(d, m)
+	log.Printf("disks currently attached to machine %v: %v", id, diskIds)
+
+	// attach the disks one by one
+	for i, diskToAttach := range disksToAttach {
+		diskId := diskToAttach["block_device_id"].(string)
+		log.Printf("Attaching the disk %v of %v (disk id: %v) to vra_machine resource %v", i+1, len(diskToAttach), diskId, d.Get("name"))
+
+		// attach the disk if it's not already attached to machine
+		if index, _ := indexOf(diskId, diskIds); index == -1 {
+			diskAttachmentSpecification := models.DiskAttachmentSpecification{
+				BlockDeviceID: withString(diskId),
+				Description:   diskToAttach["description"].(string),
+				Name:          diskToAttach["name"].(string),
+			}
+
+			attachMachineDiskOk, err := apiClient.Disk.AttachMachineDisk(disk.NewAttachMachineDiskParams().WithID(id).WithBody(&diskAttachmentSpecification))
+
+			if err != nil {
+				return err
+			}
+
+			stateChangeFunc := resource.StateChangeConf{
+				Delay:      5 * time.Second,
+				Pending:    []string{models.RequestTrackerStatusINPROGRESS},
+				Refresh:    machineStateRefreshFunc(*apiClient, *attachMachineDiskOk.Payload.ID),
+				Target:     []string{models.RequestTrackerStatusFINISHED},
+				Timeout:    d.Timeout(schema.TimeoutCreate),
+				MinTimeout: 5 * time.Second,
+			}
+
+			_, e := stateChangeFunc.WaitForState()
+			if e != nil {
+				return e
+			}
+		} else {
+			log.Printf("disk %v is already attached to machine %v, moving on to the next disk to attach", diskId, id)
+		}
+
+	}
+
+	d.SetPartial("disks")
```

Comment:
This is me pondering and I'd like your input. Is it possible and should SetPartial be called after each individual disk update (rather than at the very end)? Primarily since we're mutating the disks along the way.

Reply:
I don't think it's possible. Regarding the disks being mutated along the way, if there is an error, in the next run, it is handled thru ignoring the errors if the disk is already detached and attach only those disks that are not already attached to the machine.